kvirt/k3s/__init__.py

```diff
@@ -105,25 +112,39 @@ def create(config, plandir, cluster, overrides):
             del data['name']
         config.plan(plan, inputfile='%s/masters.yml' % plandir, overrides=data)
     firstmasterip, firstmastervmport = _ssh_credentials(k, firstmaster)[1:]
-    with open("%s/join.sh" % clusterdir, 'w') as f:
-        if api_ip is None:
-            api_ip = k.info(firstmaster)['ip']
-        joincmd = "curl -sfL https://get.k3s.io | K3S_URL=https://%s:6443 K3S_TOKEN=%s" % (api_ip, token)
-        if data['sdn'] != "flannel":
-            joincmd += " INSTALL_K3S_EXEC='--no-flannel'"
-        f.write("%s sh -\n" % joincmd)
-    source, destination = "/root/kubeconfig", "%s/auth/kubeconfig" % clusterdir
-    scpcmd = scp(firstmaster, ip=firstmasterip, user='root', source=source, destination=destination,
-                 tunnel=config.tunnel, tunnelhost=config.tunnelhost, tunnelport=config.tunnelport,
-                 tunneluser=config.tunneluser, download=True, insecure=True, vmport=firstmastervmport)
-    os.system(scpcmd)
+    ################
+    # WORKER SETUP #
+    ################
     workers = data.get('workers', 0)
     if workers > 0:
+        # Create the bash script used for the worker joining process
+        with open("%s/join.sh" % clusterdir, 'w') as f:
+            if api_ip is None:
+                api_ip = k.info(firstmaster)['ip']
+            extra_args = []
+            for component in k3s_extra_args_workers:
+                extra_args.append(component)
+            joined_k3s_extra_args_workers = " ".join(extra_args)
+            joincmd = "curl -sfL https://get.k3s.io | INSTALL_K3S_CHANNEL=%s INSTALL_K3S_VERSION=%s K3S_URL=https://%s:6443 K3S_TOKEN=%s sh -s - agent %s" % (install_k3s_channel, install_k3s_version, api_ip, token, joined_k3s_extra_args_workers)
+            f.write("apt-get -y install curl \n")
+            f.write("%s \n" % joincmd)
+            f.write("apt-get -y remove curl \n")
+        
+        # Deploy the workers
         pprint("Deploying workers")
         if 'name' in data:
             del data['name']
         os.chdir(os.path.expanduser("~/.kcli"))
-        config.plan(plan, inputfile='%s/workers.yml' % plandir, overrides=data)
+        config.plan(plan, inputfile='%s/workers.yml' % plandir, overrides=data, cluster=cluster)
```

Comment:
as commented before, in this case, data contains data['kube'], which is your cluster variable

Reply:
aaaah now I'm starting to get it 👍